transformer_lens/HookedSAE.py

```diff
@@ -0,0 +1,95 @@
+from torch import nn
+import torch
+import torch.nn.functional as F
+import einops
+from transformer_lens.hook_points import (
+    HookedRootModule,
+    HookPoint,
+)  # Hooking utilities
+from transformer_lens.HookedSAEConfig import HookedSAEConfig
+
+from typing import Union, Dict
+
+
+class HookedSAE(HookedRootModule):
+    """Hooked AutoEncoder.
+    
+    Implements a standard SAE with a TransformerLens hooks for SAE activations
+    
+    Should probably just be used for inference / analysis, not training.
+    
+    This is just a template, feel free to copy this into a notebook and change it!
+    """
+    def __init__(self, cfg: Union[HookedSAEConfig, Dict]):
+        super().__init__()
+        if isinstance(cfg, Dict):
+            cfg = HookedSAEConfig(**cfg)
+        elif isinstance(cfg, str):
+            raise ValueError(
+                "Please pass in a config dictionary or HookedSAEConfig object. If you want to load a "
+                "pretrained SAE, use HookedSAE.from_pretrained() instead."
+            )
+        self.cfg = cfg
+
+        self.W_enc = nn.Parameter(torch.nn.init.kaiming_uniform_(torch.empty(self.cfg.d_in, self.cfg.d_sae, dtype=self.cfg.dtype)))
+        self.W_dec = nn.Parameter(torch.nn.init.kaiming_uniform_(torch.empty(self.cfg.d_sae, self.cfg.d_in, dtype=self.cfg.dtype)))
+        self.b_enc = nn.Parameter(torch.zeros(self.cfg.d_sae, dtype=self.cfg.dtype))
+        self.b_dec = nn.Parameter(torch.zeros(self.cfg.d_in, dtype=self.cfg.dtype))
+        
+        self.hook_sae_input = HookPoint()
+        self.hook_sae_acts_pre = HookPoint()
+        self.hook_sae_acts_post = HookPoint() # Hook point for the SAEs hidden activations
+        self.hook_sae_recons = HookPoint()
+        self.hook_sae_error = HookPoint()
+        self.hook_sae_output = HookPoint()
+
+        self.to(self.cfg.device)
+    
+    def forward(self, input: torch.tensor):
+        """SAE Forward Pass.
+        
+        Args:
+            input: The input tensor of activations to the SAE. Shape [..., d_in]
+        """
+        self.hook_sae_input(input)
+        if input.shape[-1] == self.cfg.d_in:
+            x = input
+        else:
+            x = einops.rearrange(input, "... n_heads d_head -> ... (n_heads d_head)")
```

Comment:
```suggestion
                        # Assume this this is an attention output SAE
                        x = einops.rearrange(input, "... n_heads d_head -> ... (n_heads d_head)")
```

Reply:
Also, you should assert that the hook name `endswith` "_z" if you're going to reshape things